core/corehttp/gateway.go

```diff
@@ -12,8 +12,7 @@ func GatewayOption(writable bool) ServeOption {
 		if err != nil {
 			return err
 		}
-		mux.Handle("/ipfs/", gateway)
-		mux.Handle("/ipns/", gateway)
+		mux.Handle("/", gateway)
```

Comment:
Hmm.  In that case, this is probably best expressed as a pre-muxer request processing stage, which tries to resolve `req.Host` as an IPNS name, and if successful, rewrites the request path to `/ipns/<resolved name><path>` and allows the muxer/gateway to process the rewritten request.  I'd need to extend `ServeOption` to support registering such a stage.